rust/core-lib/src/annotations.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Management of annotations.
+
+use std::collections::HashMap;
+use serde_json::Value;
+
+use plugins::PluginId;
+use xi_rope::Interval;
+use xi_rope::spans::Spans;
+
+pub type AnnotationType = String;
+
+/// Annotation types used in core.
+pub enum CoreAnnotationType {
+    Selection,
+    Find
+}
+
+impl CoreAnnotationType {
+    pub fn as_type(&self) -> AnnotationType {
+        match self {
+            CoreAnnotationType::Selection => "selection".to_string(),
+            CoreAnnotationType::Find => "find".to_string(),
+        }
+    }
+}
+
+/// A set of annotations of a given type.
+#[derive(Clone)]
+pub struct Annotations {
+    pub items: Spans<Value>,
+    pub annotation_type: AnnotationType,
+}
+
+impl Annotations {
+    pub fn update(&mut self, interval: Interval, items: Spans<Value>) {
+        self.items.edit(interval, items);
+    }
+}
+
+/// A region of an `Annotation`.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AnnotationSlice {
+    pub annotation_type: AnnotationType,
+    /// Annotation occurrences, guaranteed non-descending start order.
+    pub ranges: Vec<(usize, usize)>,
```

Comment:
as raph mentions in some comments on the RFC, we probably don't want to use byte indexes for this, because the frontend doesn't really know about them; to the frontend everything is in visual line/col space.

Reply:
Currently, these ranges are translated into line/col when the annotations are converted to json